src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +113,252 @@ object RuleUtils {
+    // (see #145).
+    val useBucketUnion = useBucketSpec || !index.relations.head.fileFormat.equals("parquet")
+    val replacedPlan = plan transformDown {
+      case baseRelation @ LogicalRelation(
+            _ @HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+            baseOutput,
+            _,
+            _) =>
+        lazy val curFileSet = location.allFiles
+          .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
+          .toSet
+        lazy val filesAppended =
+          (curFileSet -- index.allSourceFileInfos).map(f => new Path(f.name)).toSeq
+
+        // if BucketSpec of index data isn't used (e.g., in the case of FilterIndex currently),
+        // we could read appended data from source files along with the index data.
+        val readPaths = {
+          if (useBucketUnion) {
+            index.content.files
+          } else {
+            index.content.files ++ filesAppended
+          }
+        }
+
+        val newLocation = new InMemoryFileIndex(spark, readPaths, Map(), None)
+        val relation = HadoopFsRelation(
+          newLocation,
+          new StructType(),
+          StructType(index.schema.filter(baseRelation.schema.contains(_))),
+          if (useBucketUnion) Some(index.bucketSpec) else None,
+          new ParquetFileFormat,
+          Map())(spark)
+
+        val updatedOutput =
+          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+        baseRelation.copy(relation = relation, output = updatedOutput)
+    }
+    if (useBucketUnion) {
+      // if BucketSpec of the index is used to read the index data, we need to shuffle
+      // the appended data in the same way to avoid additional shuffle of index data.
+      getComplementIndexPlan(spark, index, plan, replacedPlan)
+    } else {
+      replacedPlan
+    }
+  }
+
+  /**
+   * Get complement plan for an index with appended data.
+   *
+   * This method consists of the following steps
+   * 1) Get a plan from originalPlan by replacing data location with appended data
+   * 2) On-the-fly shuffle for the appended data, using indexedColumns & numBuckets.
+   *   - Shuffle is located before Project to utilize Push-down Filters
+   *     - Shuffle => Project => Filter => Relation
+   *   - if Project filters indexedColumns, then Shuffle should be located after the node
+   *     - Project => Shuffle => Filter => Relation
+   * 3) Bucket-aware union both indexPlan and complementPlan to avoid repartitioning index data
+   *
+   * NOTE: This method currently only supports replacement of Scan Nodes i.e. Logical relations
+   *
+   * @param spark Spark session
+   * @param index index used in replacement plan
+   * @param originalPlan original plan
+   * @param indexPlan replaced plan with index
+   * @return complementIndexPlan integrated plan of indexPlan and complementPlan
+   */
+  private def getComplementIndexPlan(
```

Comment:
This seems to be the "return" value of `getHybridScanIndexPlan` so why is this called `getComplementPlan`? What is it complementing? From what I can understand the use of complement doesn't seem appropriate. I don't have any good ideas but wanted to understand the semantic meaning here.

Reply:
I used "complement" as this additional plan (for appended files) complements the index plan. 
getCompleteIndexPlan / getComplemantaryIndexPlan / .. 🤔 

Any good suggestion would be welcomed. :)

Reply:
I saw your other renames. How about `complementIndexScanWithDataScan`? (since you are already using a verb like `transform`, I thought we could use `complement` as a word)

Reply:
I refactored the getComplemetPlan into 2 different functions - transformPlanWithAppendedFiles, shufflePlanWithIndexSpec :)